### ISSUE TITLE
Error in Building Cppinsights 

### ISSUE THREAD

**Issue by preejackie** [andreasfertig/cppinsights#56]
Hi @andreasfertig   
I'm having trouble while installing cppinsights in my machine. I have summarised in this StackOverflow [post](https://stackoverflow.com/questions/51524660/make-command-fatal-error) about my problem. I don't know exactly what was the cause of this problem. I still think this was an issue with respect to my system, could you please help me to build cppinsights :+1: 

I will close this, once I get it properly installed.
Thanks

**Comment by andreasfertig** [comment]
Hello preejackie,

it looks like llvm-config does return something the build system does not expect. Can you confirm that the following file exists:
```
/home/praveen/llvm-novice/llvm/include/clang/AST/AST.h
```

It is also suspicious that llvm_config(LLVM_SYSTEM_LIBS)=>; is empty. Usually it looks like this: 
```
llvm_config(LLVM_SYSTEM_LIBS)=>-lz -lrt -ldl -ltinfo -lpthread -lm;
```

Can you also post the output of 
```
make VERBOSE=1
```

**Comment by andreasfertig** [comment]
Hi,

`/home/praveen/llvm-novice/llvm/tools/clang/include/clang/AST` looks like the clang/llvm source path to me. At least in my installation this is the case. But not the clang/llvm install location. Did you do a ninja/make install? It looks to me that you clang/llvm setup is incomplete. That said `/home/praveen/llvm-novice/llvm/include;` also looks odd. Usually, if you have `/home/praveen/ninja/bin/clang++` then the headers are in `/home/praveen/ninja/bin/../include/` which is the case for the libs. Please ensure that you clang/llvm installation is correct.

**Comment by preejackie** [comment]
Let me install llvm and clang properly once again, and see what happens 
Hoping this time it will build.
Thanks

**Comment by preejackie** [comment]
Hi @andreasfertig   
I have built the recent version of LLVM. 

The command I used to build cppinsights is - 

cmake -GNinja ../cppinsights -DCMAKE_CXX_COMPILER=/home/praveen/live/install/bin/clang++ -DCMAKE_PREFIX_PATH=/home/praveen/live/install/bin


-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is Clang 7.0.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /home/praveen/live/install/bin/clang++
-- Check for working CXX compiler: /home/praveen/live/install/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found LLVM_CONFIG_PATH: /home/praveen/live/install/bin/llvm-config
-- llvm_config(LLVM_CXXFLAGS)=>-isystem/home/praveen/live/llvm/include -isystem/home/praveen/live/install/include  -fPIC -fvisibility-inlines-hidden -Werror=date-time -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wno-maybe-uninitialized -Wdelete-non-virtual-dtor -Wno-comment -ffunction-sections -fdata-sections -O3   -fno-exceptions -fno-rtti -D_GNU_SOURCE -D_DEBUG -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS;
-- llvm_config(LLVM_LDFLAGS)=>-L/home/praveen/live/install/lib;
-- llvm_config(LLVM_LIBS)=>-lLLVMLTO -lLLVMPasses -lLLVMObjCARCOpts -lLLVMSymbolize -lLLVMDebugInfoPDB -lLLVMDebugInfoDWARF -lLLVMFuzzMutate -lLLVMTableGen -lLLVMDlltoolDriver -lLLVMLineEditor -lLLVMOrcJIT -lLLVMCoverage -lLLVMMIRParser -lLLVMTestingSupport -lLLVMObjectYAML -lLLVMLibDriver -lLLVMOption -lgtest_main -lgtest -lLLVMWindowsManifest -lLLVMX86Disassembler -lLLVMX86AsmParser -lLLVMX86CodeGen -lLLVMGlobalISel -lLLVMSelectionDAG -lLLVMAsmPrinter -lLLVMX86Desc -lLLVMMCDisassembler -lLLVMX86Info -lLLVMX86AsmPrinter -lLLVMX86Utils -lLLVMMCJIT -lLLVMInterpreter -lLLVMExecutionEngine -lLLVMRuntimeDyld -lLLVMCodeGen -lLLVMTarget -lLLVMCoroutines -lLLVMipo -lLLVMInstrumentation -lLLVMVectorize -lLLVMScalarOpts -lLLVMLinker -lLLVMIRReader -lLLVMAsmParser -lLLVMInstCombine -lLLVMBitWriter -lLLVMAggressiveInstCombine -lLLVMTransformUtils -lLLVMAnalysis -lLLVMProfileData -lLLVMObject -lLLVMMCParser -lLLVMMC -lLLVMDebugInfoCodeView -lLLVMDebugInfoMSF -lLLVMBitReader -lLLVMCore -lLLVMBinaryFormat -lLLVMSupport -lLLVMDemangle;
-- llvm_config(LLVM_LIBDIR)=>/home/praveen/live/install/lib;
-- llvm_config(LLVM_INCLUDE_DIR)=>/home/praveen/live/llvm/include;
-- llvm_config(LLVM_SYSTEM_LIBS)=>-lz -lrt -ldl -ltinfo -lpthread -lm;
-- llvm_config(LLVM_PACKAGE_VERSION)=>7.0.0svn;
-- Git repo url: https://github.com/andreasfertig/cppinsights.git
-- Git commit hash: 5ce053d3f7bbd7dc0382385a499be13c99b67005
-- Generating version.h
-- Found clang-tidy: /home/praveen/live/install/bin/clang-tidy
-- Stripping executable
-- Found Doxygen: /usr/local/bin/doxygen (found version "1.8.14") found components:  doxygen dot 
-- Using CXX: /home/praveen/live/install/bin/clang++
-- Configuring don 
-- Generating done
-- Build files have been written to: /home/praveen/cppinsight/install

As you can see 
`llvm_config(LLVM_SYSTEM_LIBS)=>-lz -lrt -ldl -ltinfo -lpthread -lm;` is now okay. 

But when I'm running ninja, it's still finding the AST.h header file 
The output of ninja: 
1 error generated.
[6/14] Building CXX object CMakeFiles/insights.dir/CodeGenerator.cpp.o
FAILED: /home/praveen/live/install/bin/clang++   -Igenerated -I/home/praveen/live/llvm/include -I/home/praveen/cppinsight/cppinsights/. -isystem/home/praveen/live/llvm/include -isystem/home/praveen/live/install/include  -fPIC -fvisibility-inlines-hidden -Werror=date-time -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wno-maybe-uninitialized -Wdelete-non-virtual-dtor -Wno-comment -ffunction-sections -fdata-sections -O3   -fno-exceptions -fno-rtti -D_GNU_SOURCE -D_DEBUG -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -std=c++1z -Wextra -Wno-unknown-warning-option -Wold-style-cast -Werror -MD -MT CMakeFiles/insights.dir/CodeGenerator.cpp.o -MF CMakeFiles/insights.dir/CodeGenerator.cpp.o.d -o CMakeFiles/insights.dir/CodeGenerator.cpp.o -c /home/praveen/cppinsight/cppinsights/CodeGenerator.cpp
In file included from /home/praveen/cppinsight/cppinsights/CodeGenerator.cpp:8:
/home/praveen/cppinsight/cppinsights/./CodeGenerator.h:11:10: fatal error: 'clang/AST/AST.h' file not found
#include "clang/AST/AST.h"
         ^~~~~~~~~~~~~~~~~
1 error generated.
ninja: build stopped: subcommand failed.

**Comment by preejackie** [comment]
The instructions I used to build LLVM is 

cmake -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=On -DLLVM_PARALLEL_LINK_JOBS=1\
-DLLVM_ENABLE_LLD=1 -DLLVM_TARGETS_TO_BUILD="X86" ../llvm

**Comment by preejackie** [comment]
Please check whether there is a mistake in CMakeLists.txt. FYI the AST.h is available in the directory 
/home/praveen/live/llvm/tools/clang/include/clang/AST. 
I placed my clang source files in the directory /home/praveen/live/llvm/tools. 
I think that llvm-build is good.

**Comment by andreasfertig** [comment]
Hello,

this seems to be again an issue with your build system. As far as I can tell you require `libstdc++-7-dev` when using a recent clang. I just update travis to use this libstdc++ version and it works.

**Comment by preejackie** [comment]
Finally, I somehow got managed to get build correct. It is working but when I include header files in .cpp files it gives fatal error as file not found. How I can pass the path to my header files as command line arguments?

**Comment by andreasfertig** [comment]
Hello preejackie,

if I undestand your question correct, then it is the same way as for every other clang tool as well:

```
insights <YOUR_CPP_FILE> -- -I <PATH_TO_YOUR_INCLUDE_FILES>
```

**Comment by preejackie** [comment]
Hi @andreasfertig  
In short, 
I have different versions of gcc like 5.4.0, 6, 7.3, 8.1.0  installed in my systems. But by default the system uses gcc-5.4.0 installation for clang++. In clang++ command line option , we can tell clang to find different versions of gcc installed using "--gcc-toolchain" option. I tried " insights file.cpp -- -std=c++17 -extra-arg=--gcc-toolchain=/path/to/gcc-8.1/toolchain", but clang shows warning - "unused argument during compilation. 

When I tried to use insights, with this command : insights file.cpp -- -std=c++17 -stdlib=libc++ it stills throws an error : not found stddef.h, I assume this bug has been fixed in much earlier version of clang.

Clang++ --version : 
clang version 7.0.0, 

Is there any work around for this ? Or a complete re-build of cppinsights will help ? 
[clang - bug in ubuntu ](https://bugs.launchpad.net/ubuntu/+source/llvm-defaults/+bug/1242300)

As far, without including any headers in .cpp file, insights is working and I can able to see clang generated code.
